src/cpp/include/openvino/genai/vlm_pipeline.hpp

```diff
@@ -0,0 +1,483 @@
+
+    ov::Shape out_shape = embed_output_tensor.get_shape();
+    float* data = embed_output_tensor.data<float>();
+
+    embedding_dim = out_shape[out_shape.size() - 1];
+
+    //input ids embed * config.scale_emb(12)
+    for (idx = 0; idx < embed_output_tensor.get_size(); idx++) {
+        data[idx] = data[idx] * scale_emb;
+    }
+
+    //compute inputs_embedding length
+    embedding_len += 2;
+    constexpr size_t n_img_pos = 64;  // RESAMPLER query_num minicpmv-2 64, minicpmv-2.5 96
+    embedding_len += n_img_pos;
+
+    if (image_embed_slices.size() > 1) {
+        embedding_len += 1;
+        for (size_t i = 1; i < image_embed_slices.size(); ++i) {
+            for (size_t j = 0; j < image_embed_slices[i].size(); ++j) {
+                embedding_len += 2;
+                embedding_len += n_img_pos;
+
+                if (j == image_embed_slices[i].size() - 1) {
+                    embedding_len += 1;
+                }
+            }
+        }
+
+        embedding_len += 1;
+    }
+
+    ov::Tensor imgEmbedding = ov::Tensor(ov::element::f32, {1, embedding_len, embedding_dim});
+    auto imgEmbedData = imgEmbedding.data<float>();
+
+    //copy <用户> embedding info
+    memcpy(imgEmbedData, data, embed_output_tensor.get_byte_size());
+    imgEmbedData += embed_output_tensor.get_size();
+
+    //get special token embedding info
+    user_prompt = "\n<image></image><slice></slice>";
+    input_ids = tokenizer.encode(user_prompt).input_ids;
+
+    input_len = input_ids.get_size();
+
+    input_tensor = ov::Tensor(ov::element::i64, { 1, input_ids.get_size() }, input_ids.data());
+
+    embedding.set_input_tensor(input_tensor);
+    embedding.infer();
+
+    const ov::Tensor& embed_spec_tensor = embedding.get_output_tensor();
+    data = embed_spec_tensor.data<float>();
+
+    //input ids embed * config.scale_emb(12)
+    for (idx = embedding_dim; idx < embed_spec_tensor.get_size(); idx++) {
+        data[idx] = data[idx] * scale_emb;
+    }
+
+
+    //fill "<image>" embedding
+    std::copy(data + embedding_dim * 2, data + embedding_dim * 3, imgEmbedData);
+    imgEmbedData += embedding_dim;
+
+    const ov::Tensor& vision_output_tensor = image_embed_slices[0][0];
+
+    //Resampler inference with OpenVINO
+    resampler.set_tensor("x", vision_output_tensor);
+    resampler.get_tensor("tgt_size").set_shape({ 1, 1 });
+    resampler.get_tensor("tgt_size").data<int64_t>()[0] = ratio;
+    resampler.get_tensor("tgt_size").data<int64_t>()[1] = ratio;
+
+    resampler.infer();
+    const ov::Tensor& vision_embded_tensor = resampler.get_output_tensor();
+    //fill image_embed_slices[0][0]
+    std::copy_n(vision_embded_tensor.data<float>(), vision_embded_tensor.get_size(), imgEmbedData);
+    imgEmbedData += n_img_pos * embedding_dim;
+
+    //fill "</image>" embedding
+    std::copy(data + embedding_dim * 3, data + embedding_dim * 4, imgEmbedData);
+    imgEmbedData += embedding_dim;
+
+    if (image_embed_slices.size() > 1) {
```

Comment:
can we have generic loop, which will also handle case of a single slice in image?